lib/protoboeuf/codegen.rb

```diff
@@ -77,6 +77,139 @@ def resolve
       end
     end
 
+    # Generates #to_h, #as_json, and #to_json methods
+    class HashSerializationCompiler
+      include TypeHelper
+
+      attr_reader :message, :fields, :oneof_selection_fields, :generate_types
+
+      class << self
+        def result(message:, fields:, oneof_selection_fields:, generate_types:)
+          new(message:, fields:, oneof_selection_fields:, generate_types:).result
+        end
+      end
+
+      def initialize(message:, fields:, oneof_selection_fields:, generate_types:)
+        @message = message
+        @fields = fields.sort_by(&:number) # Serialize fields in their proto order
+        @oneof_selection_fields = oneof_selection_fields
+        @generate_types = generate_types
+      end
+
+      def result
+        <<~RUBY
+          #{type_signature(returns: "T::Hash[Symbol, T.untyped]")}
+          def to_h
+            result = {}
+
+            #{fields.map { |f| to_h_assign_statement_rb(f) }.join("\n")}
+
+            result
+          end
+
+          #{type_signature(params: { options: "T::Hash[T.untyped, T.untyped]" }, returns: "T::Hash[Symbol, T.untyped]")}
+          def as_json(options = {})
+            result = {}
+
+            #{fields.map { |f| as_json_assign_statement_rb(f) }.join("\n")}
+
+            result
+          end
+
+          def to_json(options = {})
+            require 'json'
+            JSON.dump(as_json(options))
+          end
+        RUBY
+      end
+
+      private
+
+      def to_h_assign_statement_rb(field)
+        key = to_h_hash_key_rb(field)
+        value = to_h_hash_value_rb(field)
+
+        if field.has_oneof_index? && !field.optional?
+          oneof_selection_field_name = oneof_selection_fields[field.oneof_index].name.dump
+          field_name = field.name.dump
+
+          "result[#{key}] = #{value} if send(:#{oneof_selection_field_name}) == :#{field_name}"
```

Comment:
Is there a reason we use `send` here?
I see it in the old code, but we (now) have (public) attribute readers for these. 🤔 
Not sure if maybe we didn't at some point.
If there is a reason, we should put a comment above this line explaining it.
If we don't, should we switch to `self.#{...}` (to still avoid any possible name clashes with local vars)?